frontend/src/components/viewers/Tensorboard.test.tsx

```diff
@@ -101,4 +110,90 @@ describe('Tensorboard', () => {
   });
+
+  it('select a version, then start a tensorboard of the corresponding version', async () => {
+    const config = { type: PlotType.TENSORBOARD, url: 'http://test/url' };
+
+    Apis.getTensorboardApp = jest.fn(() => Promise.resolve({ podAddress: '', tfVersion: '' }));
+    Apis.startTensorboardApp = jest.fn(() => Promise.resolve(''));
+    tree = mount(<TensorboardViewer configs={[config]} />);
+    await TestUtils.flushPromises();
+
+    tree
+      .find('Select')
+      .find('[role="button"]')
+      .simulate('click');
+    tree
+      .findWhere(el => el.text() === 'TensorFlow 2.0.0')
+      .hostNodes()
+      .simulate('click');
+    tree.find('BusyButton').simulate('click');
+    expect(Apis.startTensorboardApp).toHaveBeenCalledWith('http%3A%2F%2Ftest%2Furl', '2.0.0');
+  });
+
+  it('delete the tensorboard instance, confirm in the dialog, \
+    then return back to previous page', async () => {
+    Apis.getTensorboardApp = jest.fn(() =>
+      Promise.resolve({ podAddress: 'podaddress', tfVersion: '1.14.0' }),
+    );
+    const config = { type: PlotType.TENSORBOARD, url: 'http://test/url' };
+    Apis.deleteTensorboardApp = jest.fn(() => Promise.resolve(''));
+    tree = mount(<TensorboardViewer configs={[config]} />);
+    await TestUtils.flushPromises();
+    expect(!!tree.state('podAddress')).toBeTruthy();
+
+    // delete a tensorboard
+    tree.update();
+    tree
+      .find('#delete')
+      .find('Button')
+      .simulate('click');
+    tree.find('BusyButton').simulate('click');
+    expect(Apis.deleteTensorboardApp).toHaveBeenCalledWith(
+      encodeURIComponent('http://test/url'),
+      tree.state('tensorflowVersion'),
+    );
+    await TestUtils.flushPromises();
+    tree.update();
+    // the tree has returned to 'start tensorboard' page
+    expect(tree.findWhere(el => el.text() === 'Start Tensorboard').exists()).toBeTruthy();
+  });
+
+  it('show version info in delete confirming dialog, \
+    if a tensorboard instance already exists', async () => {
+    Apis.getTensorboardApp = jest.fn(() =>
+      Promise.resolve({ podAddress: 'podaddress', tfVersion: '1.14.0' }),
+    );
+    const config = { type: PlotType.TENSORBOARD, url: 'http://test/url' };
+    tree = mount(<TensorboardViewer configs={[config]} />);
+    await TestUtils.flushPromises();
+    tree.update();
+    tree
+      .find('#delete')
+      .find('Button')
+      .simulate('click');
+    expect(tree.findWhere(el => el.text() === 'Stop Tensorboard 1.14.0?').exists()).toBeTruthy();
+  });
+
+  it('click on cancel on delete tensorboard dialog, then return back to previous page', async () => {
+    Apis.getTensorboardApp = jest.fn(() =>
+      Promise.resolve({ podAddress: 'podaddress', tfVersion: '1.14.0' }),
+    );
+    const config = { type: PlotType.TENSORBOARD, url: 'http://test/url' };
+    tree = mount(<TensorboardViewer configs={[config]} />);
+    await TestUtils.flushPromises();
+    tree.update();
```

Comment:
Can you verify if these tree.update()s are necessary? Usually you shouldn't need so many of them

Reply:
Yes they are necessary. It seems that if a callback function of setState include another setState, the setState in callback would not come into effect if not update the tree.

Reply:
Got it, thanks!